File: chapters/algorithms/verlet_integration/code/c/verlet.c
#include <stdio.h>

void verlet(double pos, double acc, double dt) {
    double prev_pos, temp_pos, time;
    prev_pos = pos;
    time = 0;

    while (pos > 0) {
        time += dt;
        temp_pos = pos;
        pos = pos * 2 - prev_pos + acc * dt * dt;
        prev_pos = temp_pos;
    }

   printf("%f\n", time);
}

void stormer_verlet(double pos, double acc, double dt) {
    double prev_pos, temp_pos, time, vel;
    prev_pos = pos;
    vel = 0;
    time = 0;
    while (pos > 0) {
        time += dt;
        temp_pos = pos;
        pos = pos * 2 - prev_pos + acc * dt * dt;
        prev_pos = temp_pos;

        vel += acc * dt;
    }

   printf("%f\n", time);
}

void velocity_verlet(double pos, double acc, double dt) {
    double time, vel;
    vel = 0;
    time = 0;
    while (pos > 0) {
        time += dt;
        pos += vel * dt + 0.5 * acc * dt * dt;
        vel += acc * dt;
    }

   printf("%f\n", time);
}

int main() {
    verlet(5.0, -10, 0.01);
    stormer_verlet(5.0, -10, 0.01);
    velocity_verlet(5.0, -10, 0.01);

    return 0;
}

File: contents/contents/verlet_integration/code/golang/verlet.go
package main

import "fmt"

func verlet(pos, acc, dt float64) (time float64) {
	prevPos := pos
	time = 0

	for pos > 0 {
		time += dt
		nextPos := pos*2 - prevPos + acc*dt*dt
		prevPos, pos = pos, nextPos
	}

	return
}

func stormerVerlet(pos, acc, dt float64) (time, vel float64) {
	prevPos := pos
	time, vel = 0, 0

	for pos > 0 {
		time += dt
		vel += acc * dt
		nextPos := pos*2 - prevPos + acc*dt*dt
		prevPos, pos = pos, nextPos
	}

	return
}

func velocityVerlet(pos, acc, dt float64) (time, vel float64) {
	time, vel = 0, 0

	for pos > 0 {
		time += dt
		pos += vel*dt + .5*acc*dt*dt
		vel += acc * dt
	}

	return
}

func main() {
	time := verlet(5., -10., .01)
	fmt.Println("Verlet")
	fmt.Println("Time:", time)
	fmt.Println()

	time, vel := stormerVerlet(5., -10., .01)
	fmt.Println("Stormer-Verlet")
	fmt.Println("Time:", time)
	fmt.Println("Velocity:", vel)
	fmt.Println()

	time, vel = velocityVerlet(5., -10., .01)
	fmt.Println("Velocity Verlet")
	fmt.Println("Time:", time)
	fmt.Println("Velocity:", vel)
}
